File: pkg/utils/crypto.go
package utils

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"strings"

	"golang.org/x/crypto/bcrypt"
)

// HashPassword hash password using bcrypt (recommended) or MD5 (legacy)
func HashPassword(password string, cost int) (string, error) {
	// 默认使用bcrypt加密
	if cost == 0 {
		cost = bcrypt.DefaultCost
	}

	hashedBytes, err := bcrypt.GenerateFromPassword([]byte(password), cost)
	if err != nil {
		return "", err
	}
	return string(hashedBytes), nil
}

// HashPasswordMD5 hash password using MD5 (legacy support)
func HashPasswordMD5(password string) string {
	hash := md5.Sum([]byte(password))
	return fmt.Sprintf("%x", hash)
}

// CheckPassword verify password (supports both bcrypt and MD5)
func CheckPassword(password, hash string) bool {
	// 检查是否是bcrypt哈希（以$2a$, $2b$, $2x$, $2y$开头）
	if strings.HasPrefix(hash, "$2a$") || strings.HasPrefix(hash, "$2b$") ||
		strings.HasPrefix(hash, "$2x$") || strings.HasPrefix(hash, "$2y$") {
		err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
		return err == nil
	}

	// 兼容旧的MD5哈希
	if len(hash) == 32 {
		// 如果传入的密码已经是MD5格式，直接比较
		if len(password) == 32 {
			return password == hash
		}
		// 否则计算密码的MD5值
		passwordHash := md5.Sum([]byte(password))
		passwordHashStr := fmt.Sprintf("%x", passwordHash)
		return passwordHashStr == hash
	}

	return false
}

// ShouldUpgradePassword 检查密码是否需要升级到bcrypt
func ShouldUpgradePassword(hash string) bool {
	// 如果是MD5哈希（32位十六进制），建议升级
	return len(hash) == 32 && !strings.HasPrefix(hash, "$2")
}

// UpgradePasswordHash 将密码哈希升级到bcrypt
func UpgradePasswordHash(plainPassword string) (string, error) {
	return HashPassword(plainPassword, bcrypt.DefaultCost)
}

// GenerateSalt generate random salt
func GenerateSalt(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return base64.URLEncoding.EncodeToString(bytes), nil
}

// GenerateRandomString generate random string
func GenerateRandomString(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return base64.URLEncoding.EncodeToString(bytes)[:length], nil
}

// HashWithSalt hash with salt
func HashWithSalt(data, salt string) string {
	h := sha256.New()
	h.Write([]byte(data + salt))
	return fmt.Sprintf("%x", h.Sum(nil))
}

// GenerateSecretKey generate secret key
func GenerateSecretKey(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(bytes), nil
}

// GenerateClientSecret generate client secret
func GenerateClientSecret() (string, error) {
	return GenerateSecretKey(32)
}

// GenerateClientID generate client ID
func GenerateClientID() (string, error) {
	return GenerateRandomString(16)
}

// DecryptPassword decrypt password from frontend
func DecryptPassword(encryptedPassword string) (string, error) {
	// 加密密钥（应该与前端保持一致）
	key := []byte("eiam-platform-2024")

	// 如果密钥长度不是32字节，需要填充或截断
	if len(key) < 32 {
		// 填充到32字节
		paddedKey := make([]byte, 32)
		copy(paddedKey, key)
		key = paddedKey
	} else if len(key) > 32 {
		// 截断到32字节
		key = key[:32]
	}

	// 解码base64
	ciphertext, err := base64.StdEncoding.DecodeString(encryptedPassword)
	if err != nil {
		return "", fmt.Errorf("failed to decode base64: %v", err)
	}

	// 创建AES cipher
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("failed to create cipher: %v", err)
	}

	// 检查密文长度
	if len(ciphertext) < aes.BlockSize {
		return "", fmt.Errorf("ciphertext too short")
	}

	// 提取IV
	iv := ciphertext[:aes.BlockSize]
	ciphertext = ciphertext[aes.BlockSize:]

	// 创建CBC解密器
	mode := cipher.NewCBCDecrypter(block, iv)

	// 解密
	plaintext := make([]byte, len(ciphertext))
	mode.CryptBlocks(plaintext, ciphertext)

	// 去除填充
	plaintext = removePKCS7Padding(plaintext)

	return string(plaintext), nil
}

// removePKCS7Padding removes PKCS7 padding
func removePKCS7Padding(data []byte) []byte {
	length := len(data)
	if length == 0 {
		return data
	}

	padding := int(data[length-1])
	if padding > length {
		return data
	}

	return data[:length-padding]
}

File: frontend/src/utils/crypto.ts
import CryptoJS from 'crypto-js'

// 加密密钥（应该从环境变量或配置中获取）
const ENCRYPTION_KEY = 'eiam-platform-2024'

/**
 * 加密密码
 * @param password 原始密码
 * @returns 加密后的密码
 */
export function encryptPassword(password: string): string {
  try {
    // 使用MD5加密
    console.log('Encrypting password:', password)
    const encrypted = CryptoJS.MD5(password).toString()
    console.log('Encrypted password:', encrypted)
    return encrypted
  } catch (error) {
    console.error('Password encryption failed:', error)
    // 如果加密失败，返回原始密码（不推荐，但确保功能可用）
    return password
  }
}

/**
 * 解密密码（仅用于测试，生产环境不应解密）
 * @param encryptedPassword 加密后的密码
 * @returns 解密后的密码
 */
export function decryptPassword(encryptedPassword: string): string {
  try {
    const key = CryptoJS.enc.Utf8.parse(ENCRYPTION_KEY)
    const rawData = CryptoJS.enc.Base64.parse(encryptedPassword)
    
    // 提取IV (前16字节)
    const iv = CryptoJS.lib.WordArray.create(rawData.words.slice(0, 4))
    const ciphertext = CryptoJS.lib.WordArray.create(rawData.words.slice(4))
    
    const decrypted = CryptoJS.AES.decrypt(
      CryptoJS.lib.CipherParams.create({ ciphertext: ciphertext }),
      key,
      {
        iv: iv,
        mode: CryptoJS.mode.CBC,
        padding: CryptoJS.pad.Pkcs7
      }
    )
    
    return decrypted.toString(CryptoJS.enc.Utf8)
  } catch (error) {
    console.error('Password decryption failed:', error)
    return encryptedPassword
  }
}

/**
 * 生成随机盐值
 * @param length 盐值长度
 * @returns 随机盐值
 */
export function generateSalt(length: number = 16): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  let result = ''
  for (let i = 0; i < length; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length))
  }
  return result
}

/**
 * 哈希密码（使用盐值）
 * @param password 密码
 * @param salt 盐值
 * @returns 哈希后的密码
 */
export function hashPassword(password: string, salt: string): string {
  return CryptoJS.SHA256(password + salt).toString()
}
